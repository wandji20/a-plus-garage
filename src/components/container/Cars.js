import React, { Fragment, useEffect } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCaretLeft, faCaretRight } from '@fortawesome/free-solid-svg-icons';
import Car from '../presentation/Car';
import AddTrackButton from '../presentation/AddTrackButton';
import setFilterAction from '../../redux/actions/setFilterAction';
import { getToken } from '../../helpers/session';
import { logInUserSession } from '../../redux/actions/userAction';
import { getCars } from '../../redux/actions/carsAction';

const Cars = (props) => {
  const {
    cars, index, handleSetFilterAction, loggedIn, handleGetCars, handleLogInSession,
  } = props;

  const token = getToken('TOKEN');

  useEffect(() => {
    if (token.auth_token && !loggedIn) {
      handleLogInSession();
    }
  }, []);

  useEffect(() => {
    if (loggedIn && token.auth_token) {
      handleGetCars(token);
    }
  }, [loggedIn]);

  const handleNextChange = () => {
    if (index < cars.length - 1) {
      const newIndex = index + 1;
      handleSetFilterAction(newIndex);
    }
  };

  const handlePrevChange = () => {
    if (index > 0) {
      const newIndex = index - 1;
      handleSetFilterAction(newIndex);
    }
  };

  const car = cars[index];

  return (
    <>
      <div
        className={
            `${
              (cars.length < 1)
                ? 'd-none'
                : 'd-flex w-100 justify-content-between position-fixed car-navs'
            }`
          }
      >
        <div className="span d-inline-block">
          {
            (index > 0 && loggedIn)
              && (
                <button
                  type="button"
                  className="btn"
                  onClick={handlePrevChange}
                >
                  <FontAwesomeIcon icon={faCaretLeft} />
                </button>
              )
            }
        </div>
        <div className="span d-inline-block">
          {
              (index < cars.length - 1 && loggedIn)
              && (
                <button
                  type="button"
                  className="btn"
                  onClick={handleNextChange}
                >
                  <FontAwesomeIcon icon={faCaretRight} />
                </button>
              )
            }
        </div>
      </div>
      <div className="container remove-padding flex-column align-items-center d-flex justify-content-center">
        {
          (cars.length > 0 && loggedIn && car)
            ? <Car car={car} />
            : <AddTrackButton />
        }
      </div>
    </>
  );
};

const mapStateToProps = (state) => ({
  index: state.filterReducer.index,
  cars: state.carsReducer.cars,
  loggedIn: state.userReducer.loggedIn,
});

const mapDispatchToProps = (dispatch) => ({
  handleSetFilterAction: (index) => {
    dispatch(setFilterAction(index));
  },
  handleGetCars: () => {
    dispatch(getCars());
  },
  handleLogInSession: () => {
    dispatch(logInUserSession());
  },
});

Cars.propTypes = {
  cars: PropTypes.arrayOf(PropTypes.objectOf(PropTypes.any)).isRequired,
  index: PropTypes.number.isRequired,
  loggedIn: PropTypes.bool.isRequired,
  handleSetFilterAction: PropTypes.func.isRequired,
  handleLogInSession: PropTypes.func.isRequired,
  handleGetCars: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(Cars);
