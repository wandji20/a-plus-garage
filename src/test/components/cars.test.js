import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import store from '../../redux/store';
import Cars from '../../components/container/Cars';
import { BrowserRouter } from 'react-router-dom';

it('renders correctly', () => {
  const tree = renderer
    .create(<Provider store={store}>
        <BrowserRouter>
          <Cars />
        </BrowserRouter>
      </Provider>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
