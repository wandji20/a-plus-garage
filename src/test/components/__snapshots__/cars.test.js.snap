// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <div
    className="d-flex justify-content-between position-fixed car-navs"
  >
    <div
      className="span d-inline-block"
    />
    <div
      className="span d-inline-block"
    />
  </div>,
  <a
    href="/log_in"
    onClick={[Function]}
    style={
      Object {
        "display": "block",
        "marginTop": "200px",
      }
    }
  >
    <button
      className="btn btn-info"
      type="button"
    >
      Track Car
    </button>
  </a>,
]
`;
